test/issue8/proto.pb.go

```diff
@@ -64,6 +66,11 @@ func (m *Foo) Unmarshal(data []byte) error {
 		}
 		fieldNum := int32(wire >> 3)
 		wireType := int(wire & 0x7)
+
+		if fn := int(fieldNum - 1); fn <= 0 {
```

Comment:
I think I get it. I'm going to try to state the full logic, to see if I'm right:

1. Generate a bitmask for keeping track of which required fields are present.
1. Unmarshal. At each field, fill out the mask if that field is required.
1. After unmarshaling, check the mask for any missing fields. If found, return an error.

Is that correct? If it is, may I ask what is the benefit of fully unmarshalling, instead of bailing the moment an error is found? The latter is the approach we took in our changes to `MarshalTo`, but maybe that was naïve.